doc: add pagestorage v3 design doc (#4804)
pingcap/tiflash: docs/design/2022-04-28-architecture-of-storage-engine-pagestorage.md

The following are problems of V2 in a production environment:

- **There is a risk of data loss in the meta part**. As long as the `checksum` and `buffer size` fields in a single meta buffer are damaged at the same time, the subsequent buffers will be unavailable. This situation may occur upon a disk failure or the meta part is changed by an unexpected operation.
- **The snapshot of MVCC needs to be optimized**. First, the CPU usage occupied by the snapshot should be reduced. Second, the implementation of the MVCC structure is too complex to be understood and maintained.
- **The GC write amplification in the data part is high and the GC task is too heavy**. Because the data part is composed of append write, `compact gc` is frequently triggered. Besides, the meta part is bound to a data part by "PageFile", meaning that we have to apply `compact gc` to compact the small data in the meta part. `compact gc` means that PageStorage needs to read all the valid data from the disk, then rewrite the data into a new file. Each round of GC brings additional read and write overhead.

> Besides these three problems, we also changed the `lock` implements and the `CRC` implements.

Author: updated

## Design

![tiflash-ps-v3-architecture](./images/tiflash-ps-v3-architecture.png)

The V3 version of PageStorage is composed of three main components, `PageDirectory`, `WALStore`, and `BlobStore`.

- BlobStore: Provides an spaces management. Using the address multiplexing to manage the data part.
- PageDirectory: Provides the function of MVCC. Smaller memory usage and faster speed than V2.
- WALStore(Write Ahead Log Store): Using the write ahead log file format to manager the meta part.

### BlobStore

> BlobStore's name is earthy, but apt indicating that it mainly stores blob data. BlobStore consists of three parts

Author: changed to `BlobStore mainly stores blob data.`

1. **BlobFile**: The file stores blob data.
2. **BlobStat**: A space manager that is used to find/allocate/free spaces in BlobFile. It corresponds one-to-one with the BlobFile.
3. **BlobStats**: Manage all BlobStat. Used to schedule all write requests.

Instead of storing data with append write, BlobStore uses a data structure called SpaceMap to manage the free spaces and perform random write in BlobFile.

![tiflash-ps-v3-freemap](./images/tiflash-ps-v3-freemap.png)

> This idea comes from [Space Maps in Ext4 filesystem](https://www.kernel.org/doc/ols/2010/ols2010-pages-121-132.pdf). Each node in RB-tree has a space that is represented by (offset, size). The difference is that bitmap uses the node to record used locations, but BlobStore used the node to record free locations. Because recording free locations is better to find a free space by the data size.

Reviewer: ```suggestion
The design is inspired by [Space Maps in Ext4 filesystem](https://www.kernel.org/doc/ols/2010/ols2010-pages-121-132.pdf). Each node in RB-tree has a space that is represented by (offset, size). The difference is that bitmap uses a node to record used locations, but BlobStore uses a node to record free locations. Recording free locations is a better way to find a free space by the data size.
```

> We represent the `BlobStat` by a SpaceMap to reuse the space after being reclaimed, thereby reducing write amplification. 

Reviewer: What is the exact relationship between `BlobStat` and SpaceMap? represent here is ambiguous.
```suggestion
`BlobStat` is represented by a SpaceMap to reuse the space after being reclaimed, thereby reducing write amplification. 
```
